### PR TITLE
feat : adds openapi specifications

### PR DESCRIPTION
**Pull request by rajadilipkolli**
## Summary by CodeRabbit

- **New Features**
  - Added comprehensive RESTful API endpoints for restaurant management.
  - Introduced advanced search capabilities with multiple search strategies.
  - Implemented restaurant search functionality with geospatial and date range queries.

- **API Improvements**
  - Created new interfaces for restaurant and search operations.
  - Enhanced controller methods to support more flexible querying.
  - Simplified parameter validation across API endpoints.

- **Testing**
  - Expanded test coverage for restaurant name lookup functionality with various edge cases.
  - Added validation tests for date range and search parameter handling.
  - Introduced a new integration test for wildcard searches in boroughs.
  - Enhanced integration tests for restaurant operations.

### CODE REVIEW

File: boot-mongodb-elasticsearch/src/test/java/com/example/mongoes/web/controller/SearchControllerTest.java [hunk 1]
```diff
@@ -309,4 +299,63 @@ void whenAllParametersAreValid_thenOk() {
                     .isOk();
         }
     }
+
+    @Nested
+    class SearchDateRangeValidation {
+        @Test
+        void whenValidDateFormat_thenReturns200() {
+            given(searchService.searchDateRange("2024-01-01", "2024-12-31", 0, 10))
```
Comment (rajadilipkolli) [inline boot-mongodb-elasticsearch/src/test/java/com/example/mongoes/web/controller/SearchControllerTest.java:307]: @coderabbitai, when api call is made response is coming as 500, with below error. Please fix it

timestamp": "2025-01-17T21:22:42.975+00:00",
  "path": "/search/simple",
  "status": 500,
  "error": "Internal Server Error",
  "requestId": "5ef7fd24-7",
  "message": "Type definition error: [simple type, class org.springframework.data.elasticsearch.client.elc.ElasticsearchAggregations]",
  "trace": "org.springframework.core.codec.CodecException: Type definition error: [simple type, class org.springframework.data.elasticsearch.client.elc.ElasticsearchAggregations]\n\tat org.springframework.http.codec.json.AbstractJackson2Encoder.encodeValue(AbstractJackson2Encoder.java:257)\n\tSuppressed: The stacktrace has been enhanced by Reactor, refer to additional information below: \nError has been observed at the following site(s):\n\t*__checkpoint ⇢ Handler com.example.mongoes.web.controller.SearchController#searchSimpleQueryForBoroughAndCuisine(String, Integer, Integer) [DispatcherHandler]\n\t*__checkpoint ⇢ HTTP GET \"/search/simple?query=Manhattan%20AND%20Italian&limit=10&offset=0\"

File: boot-mongodb-elasticsearch/src/main/java/com/example/mongoes/web/api/SearchApi.java [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+                    @RequestParam
+                    @Positive
+                    Double distance,
+            @Parameter(
+                            description = "Unit of distance",
+                            example = "km",
+                            schema = @Schema(allowableValues = {"km", "mi"}))
+                    @RequestParam(defaultValue = "km", required = false)
+                    @Pattern(regexp = "^(km|mi)$", message = "Unit must be either 'km' or 'mi'")
+                    String unit);
```
[the remark this answers is not in the document]
Comment (rajadilipkolli) [reply boot-mongodb-elasticsearch/src/main/java/com/example/mongoes/web/api/SearchApi.java:433]: @coderabbitai, can you please helped me with writing junits for above suggested validations

File: boot-mongodb-elasticsearch/src/main/java/com/example/mongoes/web/api/SearchApi.java [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+            responses = {
+                @ApiResponse(
+                        responseCode = "200",
+                        description = "Successfully retrieved restaurants",
+                        content = @Content(schema = @Schema(implementation = SearchPage.class))),
+                @ApiResponse(responseCode = "400", description = "Invalid parameters provided")
+            })
+    @GetMapping("/search/terms")
+    Mono<ResponseEntity<SearchPage<Restaurant>>> searchTerms(
+            @RequestParam("query") List<String> queries,
```
[the remark this answers is not in the document]
Comment (rajadilipkolli) [reply boot-mongodb-elasticsearch/src/main/java/com/example/mongoes/web/api/SearchApi.java:159]: @coderabbitai, can you please help me to write junits for above suggestion

File: boot-mongodb-elasticsearch/src/main/java/com/example/mongoes/web/controller/RestaurantController.java [hunk 4]
```diff
@@ -2,91 +2,76 @@
...
+    public Mono<ResponseEntity<SearchPage<Restaurant>>> findAllRestaurants(int limit, int offset) {
         return restaurantService.findAllRestaurants(offset, limit).map(ResponseEntity::ok);
     }
 
-    @GetMapping("/name/{restaurantName}")
-    Mono<ResponseEntity<Restaurant>> findRestaurantByName(
-            @PathVariable @NotBlank @Size(max = 255) @Pattern(regexp = "^[a-zA-Z0-9 .-]+$")
-                    String restaurantName) {
+    @Override
+    public Mono<ResponseEntity<Restaurant>> findRestaurantByName(String restaurantName) {
```
[the remark this answers is not in the document]
Comment (rajadilipkolli) [reply boot-mongodb-elasticsearch/src/main/java/com/example/mongoes/web/controller/RestaurantController.java:38]: @coderabbitai, can you please help me to write junits for above suggestion

File: boot-mongodb-elasticsearch/src/main/java/com/example/mongoes/web/api/SearchApi.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (rajadilipkolli) [reply boot-mongodb-elasticsearch/src/main/java/com/example/mongoes/web/api/SearchApi.java:433]: @coderabbitai, below code is throwing error, please fix it .jsonPath("$.message").value(message -> 
                        assertThat(message).contains(expectedMessage));

File: boot-mongodb-elasticsearch/src/test/java/com/example/mongoes/web/controller/SearchControllerTest.java [hunk 5]
```diff
@@ -309,4 +303,321 @@
...
+                    .uri(
+                            uriBuilder ->
+                                    uriBuilder
+                                            .path("/search/terms")
+                                            .queryParam("query", "Manhattan", "")
+                                            .build())
+                    .exchange()
+                    .expectStatus()
+                    .isBadRequest();
+        }
```
[the remark this answers is not in the document]
Comment (rajadilipkolli) [reply boot-mongodb-elasticsearch/src/test/java/com/example/mongoes/web/controller/SearchControllerTest.java:568]: @coderabbitai, can you please help me with adding validation that all values in list are notBlank

File: boot-mongodb-elasticsearch/src/main/java/com/example/mongoes/web/controller/RestaurantController.java [hunk 6]
```diff
@@ -2,91 +2,76 @@
...
         return restaurantService
                 .findByRestaurantName(restaurantName)
                 .map(ResponseEntity::ok)
                 .defaultIfEmpty(ResponseEntity.notFound().build());
```
Comment (rajadilipkolli) [inline boot-mongodb-elasticsearch/src/main/java/com/example/mongoes/web/controller/RestaurantController.java:42]: @coderabbitai, how to throw `RestaurantNotFoundException` if empty, Also please let me know as per coding standards throwing from controller is better or from service

File: boot-mongodb-elasticsearch/src/main/java/com/example/mongoes/web/api/SearchApi.java [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+                        content = @Content(schema = @Schema(implementation = SearchPage.class))),
+                @ApiResponse(responseCode = "400", description = "Invalid parameters provided")
+            })
+    @GetMapping("/search/restaurant/range")
+    Mono<ResponseEntity<SearchPage<Restaurant>>> searchRestaurantIdRange(
+            @Parameter(description = "Lower limit of restaurant ID", example = "1000")
+                    @RequestParam
+                    @NotNull(message = "Lower limit is required")
+                    @Positive(message = "Lower limit must be positive")
+                    Long lowerLimit,
```
[the remark this answers is not in the document]
Comment (rajadilipkolli) [reply boot-mongodb-elasticsearch/src/main/java/com/example/mongoes/web/api/SearchApi.java:333]: @coderabbitai, can you please help me with junits for above validation
